File: mmoc/mmoc.pl
#! /usr/bin/perl

use English;

# common regular expressions
# these have to be before exift command
# (otherwise they will not get any value)
$::spaces = "\\s*";
$::comma = ",";
$::colon = ":";
$::parenO = "\\(";
$::parenC = "\\)";
$::emptyParen = "\\(\\)";
$::angleO = "<";
$::angleC = ">";
$::braceO = "\\{";
$::nameSpace = "(?:\\w+::)";
$::typeName = "\\w+";
$::pointer = "\\*";                                                                                                            
$::templateName = "\\w+";
$::plainParam = "(\\w+)";
$::boolParam = "(true|false)";
$::anyParam = "(.+)";

$::QT_MOC_PATH = find_moc ();

if (! -x $::QT_MOC_PATH) {
        print "Unable to find moc, or is not executable\n";
        if ( "MSWin32" ne "$OSNAME" ) {
                exit (1);
        }
}

chomp( $::QT_MOC_PATH );

exit main( @ARGV );

sub find_moc
{
    my $mocpath;

    if ($ENV{"QTDIR"} && -x "$ENV{\"QTDIR\"}/bin/moc")
    {
        return "$ENV{\"QTDIR\"}/bin/moc";
    }

    # windows case
    if ($ENV{"QTDIR"} && -x "$ENV{\"QTDIR\"}/bin/moc.exe")
    {
        return "$ENV{\"QTDIR\"}/bin/moc.exe";
    }

    # unix case
    if ( "MSWin32" ne "$OSNAME" )
    {

        $mocpath = `which moc 2>/dev/null`;
        if ($? == 0) {
            chomp $mocpath;
            return $mocpath;
        }

        $mocpath = `which moc-qt4 2>/dev/null`;
        if ($? == 0) {
            chomp $mocpath;
            return $mocpath;
        }
    } else {
        return "moc";
    }
}

sub main
{
    my @argv = @_;

    my @commandLineParameters = ( $::QT_MOC_PATH );

    my $filename = "";
    my $type = "";

    for ( my $i=0; $i<@argv; ++$i ) {
        if ( $argv[$i] =~ /style.h$/ ) {
            $type = "Style";
            $filename = $argv[$i];
        } elsif ( $argv[$i] =~ /model.h$/ ) {
            $type = "Model";
            $filename = $argv[$i];
        } else {
            push @commandLineParameters, $argv[$i];
        }
    }

    if ( $filename eq "" ) {
        system( @commandLineParameters );
    } else {
        push @commandLineParameters, "-f".$filename;

        if ( $type eq "Model" ) {
            runModelMoc( $filename, @commandLineParameters );
        } elsif ( $type eq "Style" ) {
            runStyleMoc( $filename, @commandLineParameters );
        }
    }

    return 0;
}

sub runStyleMoc
{
    my ($header, @arguments) = @_;

    my $commandLine = join( " ", @arguments );

    open( INF, "<$header" ) || die( "Could not open $header for reading : $!" );

    open( MOC, "|$commandLine" ) || die( "Could not run command $commandLine : $!" );

    while ( <INF> ) {
        chomp;
        my $line = $_;

        $line =~ s/\s*M_STYLE_ATTRIBUTE\s*\(\s*(\w+\:*\w*)\s*,\s*(\w+)\s*,\s*(\w+)\s*\)\s*/    Q_PROPERTY($1 $2 READ $2 WRITE set$3)/;
        $line =~ s/\s*M_STYLE_PTR_ATTRIBUTE\s*\(\s*(\w+\:*\w*\s*\*+)\s*,\s*(\w+)\s*,\s*(\w+)\s*\)\s*/    Q_PROPERTY(const $1 $2 READ $2 WRITE set$3)/;

        print MOC "$line\n";
    }

    close( MOC );

    close( INF );
}

sub runModelMoc
{
    my ($header, @arguments) = @_;

    my $commandLine = join( " ", @arguments );

    my @pattern = (
        $::spaces.
        "M_MODEL_PROPERTY".
        $::spaces.$::parenO.$::spaces.
            "(".
                "(?:".
                    $::nameSpace."?".
                    $::typeName.
                    $::spaces.
                    $::pointer."?".
                ")".
                "|".
                "(?:".
                    $::templateName.
                    $::angleO.
                        $::spaces.
                        $::typeName.
                        $::spaces.
                        $::pointer."?".
                        $::spaces.
                    $::angleC.
                ")".
            ")".
                $::spaces.$::comma.$::spaces.
            $::plainParam.
                $::spaces.$::comma.$::spaces.
            $::plainParam.
                $::spaces.$::comma.$::spaces.
            $::plainParam.
                $::spaces.$::comma.$::spaces.
            $::anyParam.
        $::spaces.$::parenC.$::spaces,

        $::spaces.
        "M_MODEL_PTR_PROPERTY".
        $::spaces.$::parenO.$::spaces.
            "(".
                "(?:".
                    $::nameSpace."?".
                    $::typeName.
                    $::spaces.
                    $::pointer."?".
                    $::spaces.
                ")".
                "|".
                "(?:".
                    $::templateName.
                    $::angleO.
                        $::spaces.
                        $::typeName.
                        $::spaces.
                        $::pointer."?".
                        $::spaces.
                    $::angleC.
                ")".
            ")".
                $::spaces.$::comma.$::spaces.
            $::plainParam.
                $::spaces.$::comma.$::spaces.
            $::plainParam.
                $::spaces.$::comma.$::spaces.
            $::boolParam.
                $::spaces.$::comma.$::spaces.
            $::anyParam.
        $::spaces.$::parenC.$::spaces,

    );

    open( INF, "<$header" ) || die( "Could not open header file for reading : $!" );

    open( MOC, "|$commandLine" ) || die( "Could not run command $commandLine : $!" );

    while ( <INF> ) {
        chomp;
        my $line = $_;

        $line =~ s/$pattern[0]/    Q_PROPERTY($1 $2 READ $2 WRITE set$3)/;
        $line =~ s/$pattern[1]/    Q_PROPERTY($1 $2 READ $2 WRITE set$3)/;

        print MOC "$line\n";
    }

    close( MOC );

    close( INF );
}

File: mmoc/mmoc.cpp
#include <QtGlobal>
#include <QFileInfo>
#include <QDir>
#include <QProcess>
#include <QtDebug>


void runModelMoc(const QString& header, const QStringList& arguments)
{
    QFile inf(header);
    if(!inf.open(QFile::ReadOnly))
        qFatal("Could not open header file for reading");

    QTextStream in(&inf);

    QProcess p;
    p.setProcessChannelMode( QProcess::ForwardedChannels );
    p.start(QString(QT_MOC_PATH), arguments, QIODevice::WriteOnly);
    if(!p.waitForStarted()) {
        qFatal("mmoc: failed to run command '%s'", QT_MOC_PATH);
    }

    QString spaces = "\\s*";
    QString comma = ",";
    QString colon = ":";
    QString parenO = "\\(";
    QString parenC = "\\)";
    QString emptyParen = "\\(\\)";
    QString angleO = "<";
    QString angleC = ">";
    QString braceO = "\\{";
    QString nameSpace = "(?:\\w+::)";
    QString typeName = "\\w+";
    QString pointer = "\\*";
    QString templateName = "\\w+";
    QString plainParam = "(\\w+)";
    QString boolParam = "(true|false)";
    QString anyParam = "(.+)";

    QRegExp propertyRegExp(
            spaces+
            "M_MODEL_PROPERTY"+
            spaces+parenO+spaces+
                "("+
                    "(?:"+
                         nameSpace+"{,1}"+
                         typeName+
                         spaces+
                         pointer+"{,1}"+
                    ")"+
                    "|"+
                    "(?:"+
                         templateName+
                         angleO+
                             spaces+
                             typeName+
                             spaces+
                             pointer+"{,1}"+
                             spaces+
                         angleC+
                     ")"+
                  ")"+
                      spaces+comma+spaces+
                  plainParam+
                      spaces+comma+spaces+
                  plainParam+
                      spaces+comma+spaces+
                  plainParam+
                      spaces+comma+spaces+
                  anyParam+
              spaces+parenC+spaces );

    QRegExp propertyPtrRegExp(
        spaces+
        "M_MODEL_PTR_PROPERTY"+
        spaces+parenO+spaces+
            "("+
                "(?:"+
                    nameSpace+"{,1}"+
                    typeName+
                    spaces+
                    pointer+"{,1}"+
                    spaces+
                ")"+
                "|"+
                "(?:"+
                    templateName+
                    angleO+
                        spaces+
                        typeName+
                        spaces+
                        pointer+"{,1}"+
                        spaces+
                    angleC+
                ")"+
            ")"+
                spaces+comma+spaces+
            plainParam+
                spaces+comma+spaces+
            plainParam+
                spaces+comma+spaces+
            boolParam+
                spaces+comma+spaces+
            anyParam+
        spaces+parenC+spaces );

    QString line;
    while(true) {
        line = in.readLine();
        if(line.isNull()) {
            break;
        }

        line.replace(propertyRegExp, "    Q_PROPERTY(\\1 \\2 READ \\2 WRITE set\\3)");
        line.replace(propertyPtrRegExp, "    Q_PROPERTY(\\1 \\2 READ \\2 WRITE set\\3)");
        p.write(QString(line + "\n").toLatin1());
    }
    p.closeWriteChannel();

    if(!p.waitForFinished()) {
        qFatal("mmoc: failed to run command '%s'", QT_MOC_PATH);
    }

    if ( p.exitCode() != 0 ) {
        qFatal("mmoc: qt moc ended with error: %i", p.exitCode() );
    }
}

void runStyleMoc(const QString& header, const QStringList& arguments)
{
    QFile inf(header);
    if(!inf.open(QFile::ReadOnly))
        qFatal("Could not open header file for reading");

    QTextStream in(&inf);

    QProcess p;
    p.setProcessChannelMode( QProcess::ForwardedChannels );
    p.start(QString(QT_MOC_PATH), arguments);
    if(!p.waitForStarted()) {
        qFatal("mmoc: failed to run command '%s'", QT_MOC_PATH);
    }


    QRegExp attributeRegExp("\\s*M_STYLE_ATTRIBUTE\\s*\\(\\s*(\\w+\\:*\\w*)\\s*,\\s*(\\w+)\\s*,\\s*(\\w+)\\s*\\)\\s*");
    QRegExp attributePtrRegExp("\\s*M_STYLE_PTR_ATTRIBUTE\\s*\\(\\s*(\\w+\\:*\\w*\\s*\\*+)\\s*,\\s*(\\w+)\\s*,\\s*(\\w+)\\s*\\)\\s*");


    QString line;
    while(true) {
        line = in.readLine();
        if(line.isNull()) {
            break;
        }

        line.replace(attributeRegExp, "    Q_PROPERTY(\\1 \\2 READ \\2 WRITE set\\3)");
        line.replace(attributePtrRegExp, "    Q_PROPERTY(const \\1 \\2 READ \\2 WRITE set\\3)");
        p.write(QString(line + "\n").toLatin1());
    }
    p.closeWriteChannel();

    if(!p.waitForFinished()) {
        qFatal("mmoc: failed to run command '%s'", QT_MOC_PATH);
    }

    if ( p.exitCode() != 0 ) {
        qFatal("mmoc: qt moc ended with error: %i", p.exitCode() );
    }
}

void runQtMoc(const QStringList& arguments)
{
    QProcess p;
    p.setProcessChannelMode( QProcess::ForwardedChannels );
    p.start(QString(QT_MOC_PATH), arguments);

    if(!p.waitForStarted()) {
        qFatal("mmoc: failed to run command '%s'", QT_MOC_PATH);
    }

    if(!p.waitForFinished()) {
        qFatal("mmoc: failed to run command '%s'", QT_MOC_PATH);
    }

    if ( p.exitCode() != 0 ) {
        qFatal("mmoc: qt moc ended with error: %i", p.exitCode() );
    }
}

enum HeaderType {
    Model,
    Style
};

int main(int argc, const char *argv[])
{
    HeaderType type=Model;
    QStringList commandLineParameters;
    QString filename;

    for(int i=1; i<argc; ++i) {
        if(QString(argv[i]).endsWith("style.h")) {
            type = Style;
            filename = argv[i];
        } else if(QString(argv[i]).endsWith("model.h")) {
            type = Model;
            filename = argv[i];
        } else {
            commandLineParameters << QString(argv[i]);
        }
    }

    if(filename.isEmpty()) {
        runQtMoc(commandLineParameters);
    } else {
        commandLineParameters << "-f" + filename;
        if(type == Model) {
            runModelMoc(filename, commandLineParameters);
        } else if(type == Style) {
            runStyleMoc(filename, commandLineParameters);
        }
    }

    return 0; //success
}

